agents/pysmurf_aggregator/stream_listener.py

```diff
@@ -0,0 +1,117 @@
+import os
+import time
+import datetime
+
+from ocs import ocs_agent, site_config
+from spt3g import core
+
+
+def _create_file_path(start_time, data_dir):
+    """Create the file path for .g3 file output.
+
+    Note: This will create directories if they don't exist already.
+
+    Arguments
+    ---------
+    start_time : float
+        Timestamp for start of data collection
+    data_dir : str
+        Top level data directory for output
+
+    Returns
+    -------
+    str
+        Full filepath of .g3 file for output
+
+    """
+    start_datetime = datetime.datetime \
+                             .fromtimestamp(start_time,
+                                            tz=datetime
+                                            .timezone.utc)
+
+    sub_dir = os.path.join(data_dir,
+                           "{:.5}".format(str(start_time)))
+
+    # Create new dir for current day
+    if not os.path.exists(sub_dir):
+        os.makedirs(sub_dir)
+
+    time_string = start_datetime.strftime("%Y-%m-%d-%H-%M-%S")
+    filename = "{}.g3".format(time_string)
+    filepath = os.path.join(sub_dir, filename)
+
+    return filepath
+
+
+class G3StreamListener:
+    def __init__(self, agent, address='localhost', port=4536):
+        self.agent = agent
+        self.log = self.agent.log
+        self.address = address
+        self.port = port
+        self.is_streaming = False
+
+    def start_stream(self, session, params=None):
+        if params is None:
+            params = {}
+
+        time_per_file = params.get("time_per_file", 60*60)  # [sec]
+        data_dir = params.get("data_dir", "data/")
+
+        self.log.info("Writing data to {}".format(data_dir))
+        self.log.info("New file every {} seconds".format(time_per_file))
+        self.log.info("Listening to {}:{}".format(self.address, self.port))
+
+        reader = core.G3Reader("tcp://{}:{}".format(self.address,
+                                                    self.port))
+        writer = None
+
+        last_meta = None
+        self.is_streaming = True
+
+        while self.is_streaming:
+            if writer is None:
+                start_time = time.time()
+                filepath = _create_file_path(start_time, data_dir)
+                writer = core.G3Writer(filename=filepath)
+
+                if last_meta is not None:
+                    writer(last_meta)
+
+            frames = reader.Process(None)
```

Comment:
I don't know if this is actually safe to do in the process loop. If the writer stops writing frames it'll freeze here, and you won't be able to cleanly exit. We should probably have a function run in a separate thread that loops while `self.is_streaming` is true and calls `reader.Process`, and adds frames to a queue. 

I think if the NetworkSender is destroyed, reader.Process will return, but its not always destroyed immediately if it's created in python. This has definitely been a problem for me in the past when testing the stream-simulator with the stream-listener.

Reply:
Right, so it does freeze if the streamer stops sending it data and you then ask the aggregator to exit.

For more context, `reader.Process()` will hang until it either receives a frame or the NetworkSender is destroyed.

Reiterating our conversation on Slack for others, I'm not sure moving this to a different thread is the right solution, as then the problem is that separate thread hangs and can't properly exit. I'm not sure the proper solution to this. I'll think about it over the weekend.